### PR TITLE
fix(gatsby-link): don't prefetch same page

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

When prefetching a url that's already being fetched can end up in a race-condition where the browser looks into it's prefetch cache.
You can end up with stale data.

How can this happen?
If you have a gatsby-link component with the same url as the current page.

This is an issue when using wordpress preview

I've added tests to make sure we test this behaviour

On the demo you see that the prefetch cached network request wins over the new request
wordpress preview bug

### CODE REVIEW

File: packages/gatsby-link/src/__tests__/__snapshots__/index.js.snap [hunk 1]
```diff
@@ -17,7 +17,7 @@ exports[`<Link /> matches basic snapshot 1`] = `
 <div>
   <a
     class="link"
-    href="/"
+    href="/active"
```
Comment (wardpeet) [inline packages/gatsby-link/src/__tests__/__snapshots__/index.js.snap:20]: The snapshot was testing the wrong thing

File: packages/gatsby-link/src/index.js [hunk 2]
```diff
@@ -108,23 +116,35 @@ class GatsbyLink extends React.Component {
     this.handleRef = this.handleRef.bind(this)
   }
 
+  _prefetch() {
+    let currentPath = window.location.pathname
+
+    // reach router should have the correct state
+    if (this.props._location && this.props._location.pathname) {
```
Comment (wardpeet) [inline packages/gatsby-link/src/index.js:123]: react location should be the correct location, window.location could be behind

**Comment by TylerBarnes** [comment]
This looks great! Thanks for doing this so quickly @wardpeet !

File: packages/gatsby-link/src/index.js [hunk 3]
```diff
@@ -181,70 +198,68 @@ class GatsbyLink extends React.Component {
...
-                }
-
-                return true
-              }}
-              {...rest}
-            />
-          ) : (
-            <a href={prefixedTo} {...rest} />
-          )
+      <Link
```
Comment (wardpeet) [inline packages/gatsby-link/src/index.js:218]: I unraveled the if and location, nothing else changed

Comment (TylerBarnes) [reply packages/gatsby-link/src/index.js:218]: For sure, looks cleaner, more dry with _prefetch() and appears that it will solve the bug 😄 Woo 🎉

**Review by TylerBarnes — APPROVED** [review]
Looks cleaner, more dry with _prefetch() and appears that it will solve the bug 😄 Woo 🎉

**Comment by TylerBarnes** [comment]
@wardpeet is there a way to fast track the release of this in the new release process? Since it's related to a P1 it'd be great to be able to ship this soon. cc @smthomas

**Comment by wardpeet** [comment]
We can hotfix it :)

**Comment by TylerBarnes** [comment]
Awesome :D

**Comment by pieh** [comment]
@wardpeet, @TylerBarnes PR backporting this to 2.27 release - https://github.com/gatsbyjs/gatsby/pull/28313

**Comment by vladar** [comment]
Published in `gatsby-link@2.6.1`, `gatsby@2.27.4`

**Comment by TylerBarnes** [comment]
Thanks @vladar !

**Pull request merged**